app/src/main/java/protect/card_locker/MainActivity.java

```diff
@@ -511,9 +490,46 @@ private void updateLoyaltyCardList(boolean updateCount) {
         }
     }
 
+    private void processBarcodeValues(Intent intent, BarcodeValues barcodeValues) {
+        if (barcodeValues.isEmpty()) {
+            return;
+        }
+        Intent newIntent = new Intent(getApplicationContext(), LoyaltyCardEditActivity.class);
+        Bundle newBundle = new Bundle();
+        newBundle.putString(LoyaltyCardEditActivity.BUNDLE_BARCODETYPE, barcodeValues.format());
+        newBundle.putString(LoyaltyCardEditActivity.BUNDLE_CARDID, barcodeValues.content());
+        Bundle inputBundle = intent.getExtras();
+        if (inputBundle != null && inputBundle.getString(LoyaltyCardEditActivity.BUNDLE_ADDGROUP) != null) {
+            newBundle.putString(LoyaltyCardEditActivity.BUNDLE_ADDGROUP, inputBundle.getString(LoyaltyCardEditActivity.BUNDLE_ADDGROUP));
+        }
+        newIntent.putExtras(newBundle);
+        startActivity(newIntent);
+    }
+
+    private void onSharedIntent(Intent intent) {
+        String receivedAction = intent.getAction();
+        String receivedType = intent.getType();
+
+        if (Intent.ACTION_SEND.equals(receivedAction)) {
+            if (receivedType.startsWith("image/")) {
+                BarcodeValues barcodeValues;
+                try {
+                    barcodeValues = Utils.parseSetBarcodeActivityResult(Utils.BARCODE_IMPORT_FROM_SHARE_INTENT, RESULT_OK, intent, this);
+                } catch (NullPointerException e) {
+                    Toast.makeText(this, R.string.errorReadingImage, Toast.LENGTH_LONG).show();
+                    return;
```

Comment:
If something goes wrong, the user probably expects to end up back at their gallery or file manager or whatever app, not Catima's main screen:

```suggestion
                    finish();
                    return;
```